src/tools/w3dviewer/colorpickerdialogclass.cpp

```diff
@@ -0,0 +1,278 @@
+    m_red = static_cast<ColorBarClass *>(GetProp(::GetDlgItem(m_hWnd, IDC_REDBAR), "CLASSPOINTER"));
+    m_green = static_cast<ColorBarClass *>(GetProp(::GetDlgItem(m_hWnd, IDC_GREENBAR), "CLASSPOINTER"));
+    m_blue = static_cast<ColorBarClass *>(GetProp(::GetDlgItem(m_hWnd, IDC_BLUEBAR), "CLASSPOINTER"));
+    m_saturation = static_cast<ColorBarClass *>(GetProp(::GetDlgItem(m_hWnd, IDC_SATURATIONBAR), "CLASSPOINTER"));
+    m_colorPicker = static_cast<ColorPickerClass *>(GetProp(::GetDlgItem(m_hWnd, IDC_COLORPICKER), "CLASSPOINTER"));
+
+    m_originalColor->SetColor(0, 0.0f, m_originalRed, m_originalGreen, m_originalBlue, 3);
+    m_colorPicker->SetColor(m_originalRed, m_originalGreen, m_originalBlue);
+
+    m_red->SetMinMax(0.0f, 255.0f);
+    m_green->SetMinMax(0.0f, 255.0f);
+    m_blue->SetMinMax(0.0f, 255.0f);
+    m_saturation->SetMinMax(0.0f, 255.0f);
+
+    m_saturation->AddColor(1, 255.0f, 255.0f, 255.0f, 255.0f, 3);
+
+    m_red->SetSliderPos(m_currentRed);
+    m_red->SetColor(0, 0.0f, 0.0f, m_currentGreen, m_currentBlue, 3);
+    m_red->SetColor(1, 255.0f, 255.0f, m_currentGreen, m_currentBlue, 3);
+
+    m_green->SetSliderPos(m_currentGreen);
+    m_green->SetColor(0, 0.0f, m_currentRed, 0.0f, m_currentBlue, 3);
+    m_green->SetColor(1, 255.0f, m_currentRed, 255.0f, m_currentBlue, 3);
+
+    m_blue->SetSliderPos(m_currentGreen);
+    m_blue->SetColor(0, 0.0f, m_currentRed, m_currentGreen, 0.0f, 3);
+    m_blue->SetColor(1, 255.0f, m_currentRed, m_currentGreen, 255.0f, 3);
+
+    m_currentColor->SetColor(0, 0.0f, m_currentRed, m_currentGreen, m_currentBlue, 3);
+
+    int cpred = 0;
+    int cpgreen = 0;
+    int cpblue = 0;
+    m_colorPicker->GetColor(&cpred, &cpgreen, &cpblue);
+    float color = m_currentRed;
+
+    if (color >= m_currentGreen) {
+        color = m_currentGreen;
+    }
+
+    if (color >= m_currentBlue) {
+        color = m_currentBlue;
+    }
+
+    m_saturation->SetSliderPos(color);
+    m_saturation->SetColor(0, 0.0f, cpred, cpgreen, cpblue, 3);
+    return TRUE;
+}
+
+int ColorPickerDialogClass::DoDialog(int *red, int *green, int *blue)
+{
+    int ret = 0;
+    ColorPickerDialogClass dlg(*red, *green, *blue, nullptr, IDD_COLORPICKER);
+
+    if (dlg.DoModal() == 1) {
+        *red = dlg.m_currentRed;
+        *green = dlg.m_currentGreen;
+        *blue = dlg.m_currentBlue;
+        ret = 1;
+    }
+
+    return ret;
+}
+
+BOOL ColorPickerDialogClass::OnCommand(WPARAM wParam, LPARAM lParam)
+{
+    if (LOWORD(wParam) >= 1020 && LOWORD(wParam) <= 1022 && HIWORD(wParam) == EN_KILLFOCUS) {
+        float red = GetDlgItemInt(1020);
+        float green = GetDlgItemInt(1021);
+        float blue = GetDlgItemInt(1022);
+        SetColor(red, green, blue, 7);
+    }
+
+    return CWnd::OnCommand(wParam, lParam);
+}
+
+BOOL ColorPickerDialogClass::OnNotify(WPARAM wParam, LPARAM lParam, LRESULT *pResult)
+{
+    NMHDR *hdr = reinterpret_cast<NMHDR *>(lParam);

```

Comment:
could simplify with ColorPickerNotify *cpn = reinterpret_cast<ColorPickerNotify *>(lParam); ?

Reply:
No because some case statements use ColorPickerNotify and others use ColorBarNotify.